[4.1] Allow User Media Field to be accessible to Registered users (Reworks #19038)

particthistle opened the pull request:
### New PR to redo #19038
Thanks to @stutteringp0et for original PR.

#### What it does
This small change allows media fields to be used in user profiles, where a user is not authorized for core.create in com_users - they can be authorized in com_media. In conjunction with a filesystem plugin, a user can be granted access to com_media a limited area of the filesystem. This change should not impact anything other than front-end media modals.

Currently the change:
- requires administrator to enable the permissions for Media component: Registered needs to have Create permissions set to Allow
![image](https://user-images.githubusercontent.com/5515866/137911317-3969af60-b3c6-4682-89e6-38b92cf564ec.png)
- provides the user with the ability to upload an image, but also select ANY media available.

A future addition / feature enhancement would be to limit user to only viewing their own media and limit uploading to just a folder for that user.

#### Summary of Changes
In `plugins/fields/media/media.php` if com_media has core.create permissions, enable the select button.
In `components/com_media/src/Dispatcher/Dispatcher.php` add a check to see if the user has permissions for core.create on com_media

#### Testing Instructions
- Create a user custom field of type media with Registered access level permissions.
- Editing your profile, the media field will have ~no button to open the media manager~ a Select button, but will not load the Media Manager when clicked because the user has no access.

- Apply this change
- Go to options for Media
- Change the permissions for the "Registered" user group, setting "create" permissions to "Allow".
- The user is now given ~the "Select"~ access to the Media Manager when they go to select an image.

#### Before Patch applied
Media manager will not load due to permissions issue
![image](https://user-images.githubusercontent.com/5515866/137912540-b7efd462-c87c-4f14-bfd3-7e731982ceb2.png)

#### Expected result
Media manager loads due to Registered Users having core.create permissions set on com_media
![image](https://user-images.githubusercontent.com/5515866/137912725-a4d19ef4-43b5-4857-983d-3033d5725537.png)

particthistle committed:
Update media.php

New PR to redo #19038

particthistle committed:
Update Dispatcher.php

Reworking #19038

brianteeman commented on plugins/fields/media/media.php:
```diff
@@ -39,7 +39,12 @@ public function onCustomFieldsPrepareDom($field, DOMElement $parent, Form $form)
 		$fieldNode->setAttribute('type', 'accessiblemedia');
-
+		
+		if (JFactory::getApplication()->getIdentity()->authorise('core.create', 'com_media'))
```
```suggestion
		if (Factory::getApplication()->getIdentity()->authorise('core.create', 'com_media'))
```

and add at the top of the file
`use Joomla\CMS\Factory;`

bembelimen committed:
Merge branch '4.1-dev' into patch-1

particthistle committed:
Update plugins/fields/media/media.php

JFactory > Factory

Refactor for Joomla 4

Co-authored-by: Brian Teeman <brian@teeman.net>

particthistle committed:
Update media.php

Added use Joomla\CMS\Factory;

particthistle committed:
Merge branch '4.1-dev' into patch-1

particthistle commented:
With #35886 doing a similar fix by removing code, which of the two PR is correct?

I've not got a moment to see if #35886 allows the User Media Field to be accessible to Registered Users, but will try and check later unless someone else has a moment to test it.

richard67 commented:
@particthistle As far as I could see, the other PR just removed obsolete conditions, and only this here allows the registered user access to the media field, so both PRs are I dependent. They could have been made together in one, but keeping it separate makes testing safer because only the relevant change is tested, here the new access aufbin the other PR that removing the obsolete stuff doesn’t break anything.

particthistle committed:
Update Dispatcher.php

Updating #35863 to incorporate changes made in #35886.

#35886 removes some code no longer required for `$asset` and `$author` variables.

#35863 then extends the changes to allow permissions for `core.create` in addition to `core.manage` to allow users with lower permissions to have create permissions assigned.

particthistle commented:
I've incorporated the changes form #35886 in advance of that being RTC.

The code removed in #35886 still leaves the problem of allowing those with core.create permissions from uploading new media. This PR resolves that based on what was initially added in #19038

particthistle committed:
Resolving 4.1-Dev conflict.

Resolving conflicts

particthistle committed:
Merge branch '4.1-dev' into patch-1

bembelimen committed:
Merge branch '4.1-dev' into patch-1

Shubhamverma2796 commented:
I have tested this item :white_check_mark: successfully on 20706fc908602d90e1695b36d775f30572df5434<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35863">issues.joomla.org/tracker/joomla-cms/35863</a>.</sub>

pritam825 commented:
I have tested this item :white_check_mark: successfully on 20706fc908602d90e1695b36d775f30572df5434<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35863">issues.joomla.org/tracker/joomla-cms/35863</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35863">issues.joomla.org/tracker/joomla-cms/35863</a>.</sub>

bembelimen committed:
Merge branch '4.1-dev' into patch-1

bembelimen commented:
Thx
